Install flow shows app details

zarvox opened the pull request:
Adapted the app details components to allow for reuse in the install flow.

If "safe", installing an app proceeds directly to the app page after successful
download and installation.  If not known safe, users are prompted with the details
of the app before confirming installation.

Fixes #15 (as of this writing, the oldest open bug!)
Fixes #1043
Fixes #1135 as part of the refactor
Closes #873

paulproteus commented on shell/client/shell.html:
```diff
@@ -597,6 +597,10 @@ <h4 class="share-with-others">Share with others</h4>
 <template name="install">
+  {{> sandstormAppInstall }}
+</template>
+<!--
```
Seems weird to leave this commented-out, fwiw!

zarvox replied:
oh, yeah, whoops.

paulproteus commented on shell/packages/sandstorm-ui-app-details/app-details-client.js:
```diff
@@ -32,7 +32,7 @@ var latestAppManifestForAppId = function (db, appId) {
 var getAppTitle = function (appDetailsHandle) {
   var pkg = latestPackageForAppId(appDetailsHandle._db, appDetailsHandle._appId);
-  return SandstormDb.appNameFromPackage(pkg);
+  return pkg && SandstormDb.appNameFromPackage(pkg) || "<unknown>";
```
The `unknown` case should never happen, right? If that's accurate, then probably we should (in the future, not necessarily now) do something to log the details of this case because if we ever do return `<unknown>` it's because of a bug somewhere.

zarvox replied:
It happens because we are doing subscription joins and the first render might not have an actual `pkg` object yet.

kentonv commented:
You requested that I write this as a comment: Let's remove the code which bypasses the interstitial when coming from the Sandstorm app market, now that the interstitial shows useful information.

kentonv commented:
UI: Mostly looks great.

The "downloading" progress view is now left-aligned, and I think it doesn't look very good that way. It really needs a design overhaul at some point, but perhaps for now it makes sense to leave it centered.

kentonv commented on shell/client/shell.html:
```diff
@@ -597,65 +597,7 @@ <h4 class="share-with-others">Share with others</h4>
-      </ol>
-    </div>
-  {{/if}}
+  {{> sandstormAppInstall }}
```
I was surprised that no parameter was being passed here, but it appears this is because the data context *is* the SandstormAppInstall object. Maybe comment on that, or write `{{> sandstormAppInstall . }}` to be explicit, or something.

zarvox replied:
Ahhh, perhaps I should do the same with the `sandstormAppDetailsPage` and `sandstormGrainListPage`

Actually, I should probably rename this `sandstormAppInstallPage` while I'm here.

zarvox commented:
Yay.  Interstitial bypass removal in progress, along with removing the "highlight" logic, since you'll never be sent through the app list that way.  (Hopefully that's not a "how do I get back here?" UX regression?  We should consider user testing this.  I expect that people will find their way back because "Apps" is fairly suggestive text.)

Re: centering: text is generally easier to scan and read when left-aligned, and the other steps will have the heading on the left, so I figured consistency and legibility justified switching to left alignment for all the steps.  Glancing over the rest of our UI, almost nothing else is centered, save the "grain loading" spinner and `/upload`, IIRC?

I could be convinced that we should change the completion fraction to be a progress bar which spans the available width, possibly in a row with the cancel button?  I could maybe be convinced to center the download percentage if there were a slightly stronger justification than "doesn't look very good that way" - I personally always thought the centered text didn't look very good.

I think the header text should remain left-aligned as it is in the rest of the views, though.  Consistency is good.

zarvox commented:
Apparently I was mistaken and the "confirm" step is always shown after all, which is great and exactly what we wanted anyway.  Hooray!

kentonv commented:
I definitely think the download dialog should be improved and the improved version could very well be left-aligned. But as-is in this PR, the feeling I get when seeing it is similar to the feeling I get when seeing raw unstyled HTML. Raw HTML is very readable, but it wouldn't be a good way to do our front page. Something something "user confidence".

> Apparently I was mistaken and the "confirm" step is always shown after all, which is great and exactly what we wanted anyway. Hooray!

If so you should be able to delete the `referredFromSandstorm()` function. :)

zarvox commented:
All right, I think I've addressed all of your comments, and the tests pass locally on my machine!

I've made the download page include a `<progress>` bar if we know the download size, and centered text otherwise.  Hopefully that makes it look less raw-HTML-y in the interim, until we can poke @neynah to sketch out something fancier.

I've also made the download update progress twice a second, which hopefully makes the download progress look a bit smoother without increasing server load too much.

kentonv commented:
Bug: If I install an unsigned SPK, I see "The app publisher has proven using PGP that they control these accounts." but there are no accounts listed. Probably the message should say "This app has no signature; its author cannot be verified." or something like that.

zarvox commented:
Fixed (along with some style improvements).

ocdtrekkie commented:
One merge button, five email notifications. :D
